relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -0,0 +1,441 @@
+//! Functionality for calculating if a trace should be processed or dropped.
+//!
+use std::convert::TryInto;
+
+use rand::{distributions::Uniform, Rng};
+use rand_pcg::Pcg32;
+use serde::{Deserialize, Serialize};
+
+use relay_common::{ProjectId, ProjectKey, Uuid};
+use relay_filter::GlobPatterns;
+
+/// A sampling rule defined by user in Organization options
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SamplingRule {
+    /// The project ids for which the sample rule applies, empty applies to all
+    #[serde(default)]
+    pub project_ids: Vec<ProjectId>,
+    /// A set of Glob patterns for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub releases: GlobPatterns,
+    /// A set of user_segments for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub user_segments: Vec<String>,
+    /// the sampling rate for trace matching this rule
+    pub sample_rate: f64,
+}
+
+/// Represents the dynamic sampling configuration available to a project.
+/// Note: This comes from the organization data
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SamplingConfig {
+    /// The sampling rules for the project
+    pub rules: Vec<SamplingRule>,
+}
+
+/// TraceContext created by the first Sentry SDK in the call chain
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TraceContext {
+    /// IID created by SDK to represent the current call flow
+    /// TODO RaduW. Should we deserialize directly into an Uuid ?
+    /// Also since trace_id and public_key are compulsory for tracing functionality should
+    /// they be an option or not ?
+    pub trace_id: Uuid,
+    /// The project key
+    pub public_key: ProjectKey,
+    /// the release
+    #[serde(default)]
+    pub release: Option<String>,
+    /// the user segment
+    #[serde(default)]
+    pub user_segment: Option<String>,
+}
+
+impl TraceContext {
+    /// Returns the decision of whether to sample or not a trace based on the configuration rules
+    /// If None then a decision can't be made either because of an invalid of missing trace context or
+    /// because no applicable sampling rule could be found.
+    pub fn should_sample(&self, config: &SamplingConfig, project_id: ProjectId) -> Option<bool> {
+        let rule = get_matching_rule(config, self, project_id)?;
+        let rate = pseudo_random_from_trace_id(self.trace_id)?;
+        Some(rate < rule.sample_rate)
+    }
```

Comment:
Can we make this function return a `bool`? This avoids the confusion in `sample_transaction_internal`. 

The intended behavior is: If no rule matches, then default sample rates should apply. The ultimate default is to retain, like in the error case.

Reply:
We can if you prefer but I think an Option<bool> is more appropriate.

A None indicates that I don't have information to make a decision (there is no matching rule). 

What we want to do in this case is already a "higher level" decision, it would be equally valid to return `true` or `false`. Of course to maintain compatibility we should include the transaction if we don't have a rule for it.

I like that `should_sample` only deals with applying the rules and  `sample_transaction_internal` deals with what to do if we don't have a rule that covers our case.

At the end of the day I'm ok with either approach.